### PR TITLE
various improvements (git, processContext, step success)

### PR DESCRIPTION
**Pull request by hohwille**
* fixed NPE I got when offline because ProcessResult.out was null
* improved GitContext - instead of retrieving all branches and remotes and looking for `*` we can simply get current branch and default remote to simplify.
* instead of checking online status 5 or more times via expensive remote operation, simply store the result
* improve step success logging that was slightly confusing

### CODE REVIEW

File: cli/src/main/java/com/devonfw/tools/ide/context/GitContextImpl.java [hunk 1]
```diff
@@ -185,89 +167,64 @@ public void clone(GitUrl gitRepoUrl, Path targetRepository) {
...
-        if (remote.startsWith("*")) {
-          String checkedOutBranch = remote.substring(remote.indexOf("[") + 1, remote.indexOf("]"));
-          remoteAndBranchName.put("remote", checkedOutBranch.substring(0, checkedOutBranch.indexOf("/")));
-          // check if current repo is behind remote and omit message
-          if (checkedOutBranch.contains(":")) {
-            remoteAndBranchName.put("branch", checkedOutBranch.substring(checkedOutBranch.indexOf("/") + 1, checkedOutBranch.indexOf(":")));
-          } else {
-            remoteAndBranchName.put("branch", checkedOutBranch.substring(checkedOutBranch.indexOf("/") + 1));
-          }
+  public String determineCurrentBranch(Path repository) {
```
Comment (jan-vcapgemini) [inline cli/src/main/java/com/devonfw/tools/ide/context/GitContextImpl.java:200]: Missing `@Override`?

File: cli/src/main/java/com/devonfw/tools/ide/context/GitContextImpl.java [hunk 2]
```diff
@@ -185,89 +167,64 @@ public void clone(GitUrl gitRepoUrl, Path targetRepository) {
...
-      return Map.ofEntries(new AbstractMap.SimpleEntry<>("remote", "unknown"), new AbstractMap.SimpleEntry<>("branch", "unknown"));
+      this.context.warning("Failed to determine current branch of git repository {}", repository);
     }
-
-    return remoteAndBranchName;
+    return null;
   }
 
-  private Optional<String[]> getLocalRemoteAndBranch(Path repositoryPath) {
+  public String determineRemote(Path repository) {
```
Comment (jan-vcapgemini) [inline cli/src/main/java/com/devonfw/tools/ide/context/GitContextImpl.java:215]: Missing `@Override`?

**Review by jan-vcapgemini — APPROVED** [review]
Thanks for your fixes, looks good to me. I've added some very small questions.

**Review by jan-vcapgemini — APPROVED** [review]

**Comment by jan-vcapgemini** [comment]
Please add bugfix to changelog.

File: cli/src/main/java/com/devonfw/tools/ide/context/GitContextImpl.java (hunk 1 above)
Comment (hohwille) [reply cli/src/main/java/com/devonfw/tools/ide/context/GitContextImpl.java:201]: Done.

**Review by jan-vcapgemini — APPROVED** [review]
I think we should add this to our changelog too. Ready to merge though.

**Comment by hohwille** [comment]
> I think we should add this to our changelog too.

Done for #650
Other changes are all internal.

**Pull request merged**